api/src/main/java/org/hawkular/inventory/api/model/RelativePath.java

```diff
@@ -0,0 +1,542 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.hawkular.inventory.api.model;
+
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Function;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonValue;
+
+/**
+ * A relative path is used in the API to refer to other entities during association. Its precise meaning is
+ * context-sensitive but the basic idea is that given a position in the graph, you want to refer to other entities that
+ * are "near" without needing to provide their full canonical path.
+ *
+ * <p>I.e. it is quite usual only associate resources and metrics from a single environment. It would be cumbersome to
+ * require the full canonical path for every metric one wants to associate with a resource. Therefore a partial path is
+ * used to refer to the metric.
+ *
+ * <p>The relative path contains one special segment type - encoded as ".." and represented using the
+ * {@link org.hawkular.inventory.api.model.RelativePath.Up} class that can be used to go up in the relative path.
+ *
+ * @author Lukas Krejci
+ * @since 0.2.0
+ */
+public final class RelativePath extends Path implements Serializable {
+
+    public static final Map<String, Class<?>> SHORT_NAME_TYPES = new HashMap<>();
+    public static final Map<Class<?>, String> SHORT_TYPE_NAMES = new HashMap<>();
+    private static final Map<Class<?>, List<Class<?>>> VALID_PROGRESSIONS = new HashMap<>();
+
+    private static final List<Class<?>> ALL_VALID_TYPES = Arrays.asList(Tenant.class, ResourceType.class,
+            MetricType.class, Environment.class, Feed.class, Metric.class, Resource.class, Up.class);
+
+    static {
+
+        SHORT_NAME_TYPES.putAll(CanonicalPath.SHORT_NAME_TYPES);
+        SHORT_NAME_TYPES.put("..", Up.class);
+
+        SHORT_TYPE_NAMES.putAll(CanonicalPath.SHORT_TYPE_NAMES);
+        SHORT_TYPE_NAMES.put(Up.class, "..");
+
+        List<Class<?>> justUp = Collections.singletonList(Up.class);
+
+        VALID_PROGRESSIONS.put(Tenant.class, Arrays.asList(Environment.class, MetricType.class, ResourceType.class,
+                Up.class));
+        VALID_PROGRESSIONS.put(Environment.class, Arrays.asList(Metric.class, Resource.class, Feed.class, Up.class));
+        VALID_PROGRESSIONS.put(Feed.class, Arrays.asList(Metric.class, Resource.class, Up.class));
+        VALID_PROGRESSIONS.put(Resource.class, Arrays.asList(Resource.class, Up.class));
+        VALID_PROGRESSIONS.put(null, Arrays.asList(Tenant.class, Relationship.class, Up.class));
```

Comment:
Moving up from a tenant leaves you in the unknown. `null` on line 73 expresses that unknown, so if your path is in the unknown "realm", you can either go down to tenants or relationships or remain in the unknown using "up".